File: skyspark-scala/src/test/scala/com/github/dkanellis/skyspark/scala/api/algorithms/bnl/DividerTest.scala
package com.github.dkanellis.skyspark.scala.api.algorithms.bnl

import com.github.dkanellis.skyspark.scala.api.algorithms.Point
import com.github.dkanellis.skyspark.scala.test_utils.{SparkAddOn, UnitSpec}
import org.mockito.Mockito._

class DividerTest extends UnitSpec with SparkAddOn {
  var flagAdder: FlagAdder = _
  var skylineComputer: SkylineComputer = _
  var divider: Divider = _

  before {
    flagAdder = mock[FlagAdder]
    skylineComputer = mock[SkylineComputer]
    divider = new Divider(flagAdder, skylineComputer)
  }

  "A set of points" should "return the local skylines with their flags" in withSpark { sc =>
    val flagPointsSeq = Seq(
      (Flag(true, true), Point(5.9, 4.6)),
      (Flag(true, true), Point(6.9, 5.6)),
      (Flag(true, false), Point(5.0, 4.1)),
      (Flag(true, false), Point(5.9, 4.0)))
    val pointsSeq = flagPointsSeq.map(_._2)
    val points = sc.parallelize(pointsSeq)
    val flagPoints = sc.parallelize(flagPointsSeq)
    when(flagAdder.addFlags(points)).thenReturn(flagPoints)
    val expectedLocalSkylinesWithFlagsSeq = Seq(
      (Flag(true, true), Point(5.9, 4.6)),
      (Flag(true, false), Point(5.0, 4.1)),
      (Flag(true, false), Point(5.9, 4.0)))
    val expectedLocalSkylinesWithFlags = sc.parallelize(expectedLocalSkylinesWithFlagsSeq)
    when(skylineComputer.computeLocalSkylines(flagPoints)).thenReturn(expectedLocalSkylinesWithFlags)

    val actualLocalSkylinesWithFlags = divider.divide(points).collect

    actualLocalSkylinesWithFlags should contain theSameElementsAs expectedLocalSkylinesWithFlagsSeq
  }
}

File: skyspark-java/src/test/java/com/akikanellis/skyspark/api/algorithms/bnl/DividerTest.java
package com.akikanellis.skyspark.api.algorithms.bnl;

import com.akikanellis.skyspark.api.algorithms.Point;
import com.akikanellis.skyspark.api.test_utils.SparkContextRule;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;
import scala.Tuple2;

import java.util.Arrays;

import static com.akikanellis.skyspark.api.test_utils.assertions.MoreAssertions.assertThatP;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class DividerTest {
    @Rule public SparkContextRule sc = new SparkContextRule();

    @Mock private FlagAdder flagAdder;
    @Mock private LocalSkylineCalculator localSkylineCalculator;
    private Divider divider;

    @Before public void beforeEach() { divider = new Divider(flagAdder, localSkylineCalculator); }

    @Test public void producingLocalSkylines_returnsCorrectSkylines() {
        JavaRDD<Point> points = sc.parallelize(Arrays.asList(
                new Point(5.9, 4.6),
                new Point(6.9, 5.6),
                new Point(5.0, 4.1),
                new Point(5.9, 4.0))
        );
        JavaPairRDD<Flag, Point> flagPoints = sc.parallelizePairs(Arrays.asList(
                new Tuple2<>(new Flag(true, true), new Point(5.9, 4.6)),
                new Tuple2<>(new Flag(true, true), new Point(6.9, 5.6)),
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0)))
        );
        when(flagAdder.addFlags(points)).thenReturn(flagPoints);
        JavaPairRDD<Flag, Point> expectedLocalSkylinesWithFlags = sc.parallelizePairs(Arrays.asList(
                new Tuple2<>(new Flag(true, true), new Point(5.9, 4.6)),
                new Tuple2<>(new Flag(true, false), new Point(5.0, 4.1)),
                new Tuple2<>(new Flag(true, false), new Point(5.9, 4.0)))
        );
        when(localSkylineCalculator.computeLocalSkylines(flagPoints)).thenReturn(expectedLocalSkylinesWithFlags);

        JavaPairRDD<Flag, Point> actualLocalSkylinesWithFlags = divider.divide(points);

        assertThatP(actualLocalSkylinesWithFlags).containsOnlyElementsOf(expectedLocalSkylinesWithFlags);
    }
}